src/libAtomVM/memory.h

```diff
@@ -45,6 +47,131 @@ enum MemoryGCResult
     MEMORY_GC_DENIED_ALLOCATION = 2
 };
 
+// mso_list is the first term of heap
+#define MSO_LIST_TERM_INDEX 0
+
+struct HeapFragment;
+typedef struct HeapFragment HeapFragment;
+
+struct HeapFragment
+{
+    HeapFragment *next;
+    term *heap_end;
+    term storage[];
+};
+
+struct Heap
+{
+    HeapFragment *next;
+    term *heap_start;
+    term *heap_ptr;
+    term *heap_end;
+};
+typedef struct Heap Heap;
+
+#define HEAP_START_TO_FRAGMENT(heap_start) \
+    ((HeapFragment *) (((char *) (heap_start)) - ((unsigned long) &((HeapFragment *) 0)->heap_end)))
+
+#define BEGIN_WITH_STACK_HEAP(size, name)    \
+    {                                        \
+        term heap_storage[size];             \
+        Heap name;                           \
+        name.heap_start = heap_storage;      \
+        name.heap_end = heap_storage + size; \
+        memory_init_heap_fragment(&name);
+
+#define END_WITH_STACK_HEAP(name)               \
+    if (name.next) {                            \
+        memory_deinit_heap_fragment(name.next); \
+    }                                           \
+    }
+
```

Comment:
I think it would be useful to also have macros for creating and destroying heap fragments on the (C) heap, as some ports may need to create large and complicated terms, and putting them on the stack could put the VM in danger of crashing.

Reply:
Isn't memory_ensure_free_opt(... MEMORY_NO_GC) suitable for this?

Reply:
Actually such ports should call `memory_init_heap` and `memory_deinit_heap` instead of using the macro.

Reply:
(see my comment)